### PR TITLE
Feature/scalar-value-interval-ops

### PR DESCRIPTION
**Pull request by berkaysynnada**
# Which issue does this PR close?

Closes #5412.

# Rationale for this change

Currently we cannot compare and make addition and subtraction operation with two ScalarValue interval types, `IntervalYearMonth`, `IntervalDayTime`, `IntervalMonthDayNano`. With this PR, we can do those between the same variants and different variants.

# What changes are included in this PR?

`impl PartialOrd for ScalarValue` and `impl PartialEq for ScalarValue` are extended to handle interval types. However, to be able to compare months and days, we need to assume a month equals to 30 days (postgre behaves in the same way). For the arithmetic operations, `macro_rules! impl_op` is extended with some helper functions.

# Are these changes tested?

Yes. Some tests are written to cover all variations of operations for different types.

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/common/src/scalar.rs [hunk 1]
```diff
@@ -554,6 +669,110 @@ macro_rules! impl_op {
...
+        )))
+    } else {
+        Ok(ScalarValue::IntervalMonthDayNano(Some(
+            IntervalMonthDayNanoType::make_value(
+                *lhs * sign,
+                rhs.0,
+                (rhs.1) as i64 * 1_000_000,
+            ),
+        )))
+    }
```
Comment (mustafasrepo) [inline datafusion/common/src/scalar.rs:723]: replace body with 
```rust
    let (mut days, millis) = IntervalDayTimeType::to_parts(*rhs);
    let mut nanos = millis as i64 * 1_000_000;
    let mut lhs = *lhs;

    if commute {
        lhs = lhs*sign;
    } else {
        days = days * sign;
        nanos = nanos * sign as i64;
    };
    Ok(ScalarValue::IntervalMonthDayNano(Some(
        IntervalMonthDayNanoType::make_value(
            lhs,
            days,
            nanos,
        ),
    )))
```

File: datafusion/common/src/scalar.rs [hunk 2]
```diff
@@ -554,6 +669,110 @@ macro_rules! impl_op {
     };
 }
 
+#[inline]
+fn op_ym(lhs: &i32, rhs: &i32, sign: i32) -> Result<ScalarValue, DataFusionError> {
```
Comment (mustafasrepo) [inline datafusion/common/src/scalar.rs:673]: remove references from `lhs`, and `rhs`. Do this for all of the functions that takes &i64, &i32 etc.

File: datafusion/common/src/scalar.rs [hunk 3]
```diff
@@ -554,6 +669,110 @@ macro_rules! impl_op {
...
+                *lhs + rhs.0 * sign,
+                rhs.1 * sign,
+                rhs.2 * sign as i64,
+            ),
+        )))
+    } else {
+        Ok(ScalarValue::IntervalMonthDayNano(Some(
+            IntervalMonthDayNanoType::make_value(*lhs * sign + rhs.0, rhs.1, rhs.2),
+        )))
+    }
```
Comment (mustafasrepo) [inline datafusion/common/src/scalar.rs:745]: replace body with
```rust
    let (mut months, mut days, mut nanos) = IntervalMonthDayNanoType::to_parts(*rhs);
    if commute {
        months = months + (*lhs * sign);
    } else {
        months = *lhs + (months * sign);
        days *= sign;
        nanos *= sign as i64;
    }
    Ok(ScalarValue::IntervalMonthDayNano(Some(
        IntervalMonthDayNanoType::make_value(months, days, nanos),
    )))
```

File: datafusion/common/src/scalar.rs [hunk 4]
```diff
@@ -554,6 +669,110 @@ macro_rules! impl_op {
...
+}
+#[inline]
+fn op_dt_mdn(
+    lhs: &i64,
+    rhs: &i128,
+    sign: i32,
+    commute: bool,
+) -> Result<ScalarValue, DataFusionError> {
+    let lhs = IntervalDayTimeType::to_parts(*lhs);
```
Comment (mustafasrepo) [inline datafusion/common/src/scalar.rs:754]: Do similar change with above examples

**Comment by mustafasrepo** [comment]
As far as I am concerned this PR is ready to send upstream.

**Comment by ozankabak** [comment]
Closing this as we will send a one joint PR upstream. See the rationale in the other PR.

**Pull request closed without merging**